sh.c

```diff
@@ -1778,8 +1782,60 @@ bool comp_switch_block_statement(ast node, bool start_in_tail_position) {
     comp_statement(node, false);
     return true;
   } else if (get_op(node) == CASE_KW || get_op(node) == DEFAULT_KW) {
-    fatal_error("comp_statement: case must be at the beginning of a switch block, and each block must end with a break or return statement");
+    fatal_error("comp_switch_block_statement: case must be at the beginning of a switch block, and each block must end with a break or return statement");
+    return false;
+  } else if (get_op(node) == '{') {
+    statement = get_child(node, 0);
+    while(statement != 0) {
+      if (comp_switch_block_statement(statement, false, start_in_tail_position)) {
+        return true;
+      }
+      statement = get_child(node, 1);
+    }
     return false;
+  } else if (get_op(node) == IF_KW) {
+    append_glo_decl(string_concat3(
+          wrap_str_lit(else_if ? "elif " : "if "),
+          comp_rvalue(get_child(node, 0), else_if ? RVALUE_CTX_TEST_ELSEIF : RVALUE_CTX_TEST),
+          wrap_str_lit(" ; then")
+    ));
+
+    nest_level += 1;
+    if (get_child(node, 1) != 0) {
+      if(get_op(get_child(node, 1)) == BREAK_KW) {
+        append_glo_decl(wrap_char(':'));
+        termination_lhs = true; 
+      } else {
+        termination_lhs = comp_switch_block_statement(get_child(node, 1), false, start_in_tail_position); 
+      }
+    } else {
+      append_glo_decl(wrap_char(':'));
+    }
+    nest_level -= 1;
+
+    // else
+    if (get_child(node, 2) != 0) {
+      // Compile sequence of if else if using elif
+      if (get_op(get_child(node, 2)) == IF_KW) {
+        termination_rhs = comp_switch_block_statement(get_child(node, 2), true, start_in_tail_position); // comp_statement with else_if == true emits elif
+      } else {
+        append_glo_decl(wrap_str_lit("else"));
+        nest_level += 1;
+        if (get_op(get_child(node, 2)) == BREAK_KW)  {
+          append_glo_decl(wrap_char(':'));
+          termination_rhs = true;
+        } else {
+          termination_rhs = comp_switch_block_statement(get_child(node, 2), false, start_in_tail_position);
+        }
+        nest_level -= 1;
+      }
+    }
+
+    if (!else_if) {
+       append_glo_decl(wrap_str_lit("fi"));
+    }
+
+    return termination_lhs && termination_rhs;
```

Comment:
I think in the case `termination_lhs == true && !termination_rhs` we want to throw an error because we found an early return and that's not well supported so better fail to compile than to generate invalid code.

Reply:
Yep, fine for me. The actual case would be `termination_lhs ^ termination_rhs` AFAICT